File: Algorithms/Search_Algorithms/Binary_Search/Koko_Eating_Bananas/Koko_Eating_Bananas.py
# -*- coding: utf-8 -*-
"""
Created on Wed Oct 14 23:34:35 2020

@author: Abhinav Tiwari

Solution to question as per QUESTION.md's
 
    Koko eating bananas
    finds eating rate K bananas/hr 
    So, that she can finish the pile 
    before the guards arrive.
"""
import math


def min_rate(pile,hrs):
    res = 0
    for i in range(len(pile)):
        res += math.ceil(pile[i]/hrs)
    return res    
        
def banana_per_hour(pile,hrs):
    
    if len(pile)==1:
        return math.ceil(pile[0]/hrs)
    
    lo = 1
    hi = max(pile)
    
    value = -986765
    while(lo<hi):
        mid = lo + (hi-lo)//2
        value = min_rate(pile, mid)
        if value<=hrs:
            hi = mid
        else:
            lo = mid + 1
    
    return lo
       
    
    
def main():
    bananas = []
    print("Enter the bananas in pile one by one",end = "\n")
    print("Enter any negative no to quit",end ='\n')
    n = int(input())
    while(n>0):
        bananas.append(n)
        n = int(input())
    H = int(input("Enter hours till the guards arrive:    ")) 
    K = banana_per_hour(bananas, H)
    print("For her to eat all the bananas till the guards arrive, she must eat @ ",K,"bananas per hour")
main()

File: Algorithms/Searching/Binary_Search/Koko_Eating_Bananas/Koko_Eating_Bananas.cpp
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;
int computeMinHours(vector<int> &piles, int speed)
{
    int n = piles.size();
    int res = 0;
    int i = 0;
    for (i = 0; i < n; i++)
        res += ceil((float)piles[i] / speed);
    return res;
}
int computeMax(vector<int> &piles)
{
    int n = piles.size();
    int maxval = INT_MIN;
    for (int i = 0; i < n; i++)
        maxval = max(maxval, piles[i]);
    return maxval;
}
int minEatingSpeed(vector<int> &piles, int H)
{
    int n = piles.size();
    if (n == 1)
    {
        double b = (double)piles[0];
        int temp = ceil(b / H);
        return temp;
    }
    int lo = 1;
    int hi = computeMax(piles);
    int mid;
    int intval;
    while (lo < hi)
    {
        mid = lo + (hi - lo) / 2;
        intval = computeMinHours(piles, mid);
        if (intval <= H)
        {
            hi = mid;
        }
        else
        {
            lo = mid + 1;
        }
    }
    return lo;
}

int main()
{
    vector<int> piles;
    int H;
    cout << "Enter contents of each pile (enter any negative number to stop)\n";
    while (true)
    {
        int x;
        cout << "Enter content of pile: ";
        cin >> x;
        if (x < 0)
            break;
        piles.push_back(x);
    }
    cout << "Enter number of hours for guards to return \n";
    cin >> H;
    int res = minEatingSpeed(piles, H);
    cout << "The minimum eating speed should be " << res << " bananas per hour";
    return 0;
}